docs/blog/posts/write-code-to-refactor.md

```diff
@@ -0,0 +1,402 @@
+
+### Get a small pipeline first
+
+When working with longer training jobs, it's tempting to try doing a YOLO run at the start and seeing if things work. However, as your pipeline gets more and more complex, you might not be able to detect errors in your logic easily. A good way to circumvent this is to work first with a smaller model and a smaller dataset to ensure that your entire pipeline works from start to finish. 
+
+I typically work with a training dataset of 1000-2000 values when I'm writing these scripts. My goal is to ideally have something that can be ran in < 60s at most so that I can check for any quick implementation issues. This has caught many different bugs in my codebase because we can iterate and experiment quickly.
+
+### Using Checkpoints
+
+It's not uncommon for models to fail midway through a long training run either due to a timeout issue or insufficient memory capacity inside the CUDA GPU that you're using to train
+
+![Solving the “RuntimeError: CUDA Out of memory” error | by Nitin Kishore |  Medium](https://miro.medium.com/v2/resize:fit:1400/1*enMsxkgJ1eb9XvtWju5V8Q.png)
+
+Therefore, you'll need to implement some form of checkpoints so that if training fails, you can just resume it from an earlier version. This has been a lifesaver in many circumstances and many standard libraries should support it out of the box.
+
+### Use Generators
+
+Generators allow you to get significant performance improvements. Let's take the following two code examples
+
+```python
+def get_all_keys(data):
+  return [row["key"] for row in data]
+
+def get_all_keys_with_generator(data):
+	for row in data:
+    yield row["key"]
+```
+
+In the second function, we use a generator to provide access to the data which allows us to load the data on-demand. This helps us to speed-up the execution of our program since now we can load the data on demand. We can see how a generator helps us by running a simple benchmark as seen below
+
+```python
+def benchmark(n_rows: int):
+    data = [{"key": 24} for _ in range(n_rows)]
+    start = time.perf_counter()
+    keys = get_all_keys(data)
+    ttl = sum(keys)
+    end = time.perf_counter()
+
+    generator_start = time.perf_counter()
+    gen_keys = get_all_keys_with_generator(data)
+    gen_ttl = sum(gen_keys)
+    generator_end = time.perf_counter()
+
+    assert gen_ttl == ttl
+
+    return end - start, generator_end - generator_start
+
+
+for sz in [100, 1000, 10000, 100000, 1000000, 10000000, 100000000]:
+    list_implementation, generator_implementation = benchmark(sz)
+    print(
+        f"Size: {sz:<15,.0f} - Generator: {generator_implementation:10,.10f}, List:{list_implementation:10,.10f}. Generator is {round(list_implementation/generator_implementation,3)} times faster"
+    )
+```
+
+We can see the output of the script above in the code block below.
+
+```bash
+Size: 100             - Generator: 0.0000052080, List:0.0000053750. Generator is 1.032 times faster
```

Comment:
It's not about being faster per se, it's about the simplicity of the code. 

And you shouldn't be writing your own benchmark code, you should use something like` timeit` You guys are not capturing any variants.